internal/server/singleprocess/service_logs.go

```diff
@@ -305,14 +313,161 @@ func (s *service) GetLogStream(
+func (s *service) sendInstanceLogs(
+	ctx context.Context,
+	log hclog.Logger,
+	sender batchSender,
+	instanceFunc func(ws memdb.WatchSet) ([]*streamRec, error),
+	backlog int32,
+) error {
 	// We keep track of what instances we already have readers for here.
 	var instanceSetLock sync.Mutex
 	instanceSet := make(map[string]struct{})
 
+	// For values returned by LogMerge, we want to be able to map back to
+	// the stream that the reader was for, so we can include the instanceid.
+	// This map lets us do that.
+	readerToInstance := make(map[*logbuffer.Reader]*streamRec)
+
+	// Step 1: we use log merge to read all the known entries from existing
+	// instances. This will never block, it will just weave the log entries
+	// that are already known together and then send them back to the client.
+
+	// Get all current records
+	ws := memdb.NewWatchSet()
+	records, err := instanceFunc(ws)
+	if err != nil {
+		return err
+	}
+	log.Trace("instances loaded", "len", len(records))
+
+	var readers []logbuffer.MergeReader
+
+	for _, record := range records {
+		r := record.LogBuffer.Reader(backlog)
+		readerToInstance[r] = record
+
+		readers = append(readers, r)
+	}
+
+	lm := logbuffer.NewMerger(readers...)
+
+	lines := make([]*pb.LogBatch_Entry, maxEntriesPerRead)
+
+	// Read out all the log entries from LogMerge. This never blocks waiting
+	// for new log entries, it will simply let each reader output all known
+	// entries and then loop exits.
+	for {
+		entries, err := lm.Read(len(lines))
+		if err != nil {
+			return err
+		}
+
+		// When there are no more buffered entries to read, means
+		// we'll switch to the on-demand reading logic below.
+		if len(entries) == 0 {
+			break
+		}
+
+		log.Trace("sending known log data", "entries", len(entries))
+
+		var (
+			prev *streamRec
+			idx  int
+		)
+
+		// We batch up the lines so long as the previous line is from the
+		// same instance as the current one. When we detect a change, we flush
+		// lines and begin buffering again.
+		for _, v := range entries {
+			rec := readerToInstance[v.Reader.(*logbuffer.Reader)]
+
+			if prev != nil && prev != rec {
+				// Flush current lines that were all the same instance
+				sender.Send(&pb.LogBatch{
+					DeploymentId: prev.DeploymentId,
+					InstanceId:   prev.InstanceId,
+					Lines:        lines[:idx],
+				})
+
+				idx = 0
+			}
+
+			val := v.Value().(*pb.LogBatch_Entry)
+			lines[idx] = val
+
+			prev = rec
+			idx++
+		}
+
+		// Flush any unsent lines
+		if idx > 0 {
+			// Flush current lines that were all the same instance
+			sender.Send(&pb.LogBatch{
+				DeploymentId: prev.DeploymentId,
+				InstanceId:   prev.InstanceId,
+				Lines:        lines[:idx],
+			})
+		}
+	}
+
+	// Step 2: startup background forwarders for all the readers we spawned above.
+
+	// We lock around this section because if one of the launched goroutines exits
+	// very quickly, we might still be adding the others when it does exit.
+	{
+		instanceSetLock.Lock()
```

Comment:
This is the only thing that bothers me about the stream changes, is that we _almost kind of_ duplicate this logic twice (here and just below). I didn't think about it too hard, but is there not some common logic we can extract into a func to setup the log forwarding or something and call that in both places? The juggling of the `instanceSetLock` in particular in two places kind of makes me nervous.

Reply:
So I refactored this code a few different ways to try and deal with that lock and their were all either confusing and didn't generally improve the understanding. Because of the bifurcated nature of the types of readers, this ended up being the version that, while did require some mutex manipulation duplication, was the cleanest to read and understand.

Reply:
Fair enough!